File: Data Structures & Algorithms/meeting-schedule-ii/submission-1.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        intervals.sort(key=lambda x: ( x.start, x.end ) )
        # store the latest ending time of each room
        rooms = []
        for interval in intervals:
            start = interval.start
            end = interval.end

            found = False
            for i in range(len(rooms)):
                if start >= rooms[i]:
                    rooms[i] = end
                    found = True
                    break
            if not found:
                rooms.append(end)

        return len(rooms)

File: Data Structures & Algorithms/meeting-schedule/submission-1.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    bool canAttendMeetings(vector<Interval>& intervals) {
        sort( intervals.begin(), intervals.end(), [](const Interval& a, const Interval& b){
            return a.start < b.start;
        });
        int currentEnd = 0;
        for( auto interval : intervals ){
            if( interval.start >= currentEnd ){
                currentEnd = interval.end;
            }
            else
                return false;
        }
        return true;
    }
};
